### PR TITLE
Feature/ln/backend

### PR DESCRIPTION
**Pull request by lukasNwr**
Definitely needs proper checking, because I don't trust myself with many things that I tried to do :D

### CODE REVIEW

File: Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ * Environment variables needed
+ * 
+ * SOLVER_MANAGER_ADDRESS   - Local network address of Solver manager
+ * DB_SERVICE_ADDRESS       - Address of db service 
+ * BUCKER_HANDLER_ADDRESS   - Address of bucket handler
+ */
+
+@Service
+public class BackendService {
+    private String solverManagerAddress = System.getenv("SOLVER_MANAGER_ADDRESS");
```
Comment (Maje419) [inline Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java:36]: Each address will consist of two different environment variables, the service name and the port. These are fetched from the "green-peace.yaml" file within the "Deployments" folder when defining a new deployment.
```suggestion
    private String solverManagerAddress = "http://" + System.getenv("SOLVER_MANAGER_SERVICE") + ":" + System.getenv("SOLVER_MANAGER_PORT");
```

File: Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    }
+
+    public String getAllLegalSolvers() throws JsonProcessingException {
+        String url = dbServiceAddress + "/solvers";
+        Solver[] solvers = restTemplate.getForObject(url, Solver[].class);
+        String jsonData = objectMapper.writeValueAsString(solvers);
+
+        return jsonData;
+
```
Comment (Maje419) [inline Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java:60]:
```suggestion
```

File: Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    ObjectMapper objectMapper = new ObjectMapper();
+
+    public BackendService() {
+    }
+
+    public void postSolversToSolverManager(SolveRequest request) {
+        String url = solverManagerAddress + "/New";
+        restTemplate.postForEntity(url, request, SolveRequest.class);
+
```
Comment (Maje419) [inline Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java:51]:
```suggestion
```

File: Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    public String cancelSolver(String solverName, CancelSolverRequest cancelSolverRequest) {
+        String url = solverManagerAddress + "/cancel/user";
+        restTemplate.postForEntity(url, cancelSolverRequest,
+                CancelTaskRequest.class);
+        return solverName;
+
+    }
+
```
Comment (Maje419) [inline Backend/src/main/java/com/oops/backend/Backend/services/BackendService.java:198]:
```suggestion
```

File: Backend/src/main/resources/application.properties [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (Maje419) [inline Backend/src/main/resources/application.properties:1]: Is this file strictly needed even if it's empty?

File: Backend/src/test/java/com/oops/backend/Backend/BackendApplicationTests.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.oops.backend.Backend.models.Solver;
+import com.oops.backend.Backend.services.BackendService;
+
+@SpringBootTest
+class BackendApplicationTests {
+
+	@Test
+	void contextLoads() {
```
Comment (Maje419) [inline Backend/src/test/java/com/oops/backend/Backend/BackendApplicationTests.java:25]: Still more thorough than the BucketHandler tests lol xD

**Review by Maje419 — CHANGES_REQUESTED** [review]
Very good implementation, clear code.

I'll approve it once the adresses are correct

File: Backend/src/main/java/com/oops/backend/Backend/requests/SolversToUseBody.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package com.oops.backend.Backend.requests;
+
+public class SolversToUseBody {
+    private String NumberVCPU;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/requests/SolversToUseBody.java:4]: I made a mistake in the API spec, we also need a field called SolverName.

File: Backend/src/main/java/com/oops/backend/Backend/requests/CancelSolverRequest.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package com.oops.backend.Backend.requests;
+
+public class CancelSolverRequest {
+    private String problemID;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/requests/CancelSolverRequest.java:4]: The API spec was slightly wrong, it is changed now, should be "ProblemID and UserID.

File: Backend/src/main/java/com/oops/backend/Backend/requests/CancelTaskRequest.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package com.oops.backend.Backend.requests;
+
+public class CancelTaskRequest {
+    private String userId;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/requests/CancelTaskRequest.java:4]: api spec changed slightly, should be "UserID

File: Backend/src/main/java/com/oops/backend/Backend/requests/CancelUserTasksRequest.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.oops.backend.Backend.requests;
+
+public class CancelUserTasksRequest {
+    private String userID;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/requests/CancelUserTasksRequest.java:4]: API spec changed, should be "UserID"

File: Backend/src/main/java/com/oops/backend/Backend/models/SolverBody.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
+package com.oops.backend.Backend.models;
+
+public class SolverBody {
+    private String solverName;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/models/SolverBody.java:4]: I dont now if this is relevant with converting from Json to model, but api spec changed to have the first letter capitalized.

File: Backend/src/main/java/com/oops/backend/Backend/controllers/BackendController.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    @GetMapping("/DataInstance/{ProblemID}")
+    public ResponseEntity<String> getDZNData(@PathVariable String ProblemID) {
+        String dznData = backendService.getDznData(ProblemID);
+        return new ResponseEntity<String>(dznData, HttpStatus.OK);
+    }
+
+    @GetMapping("/Solver")
+    public ResponseEntity<String> getSolvers() throws JsonProcessingException {
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/controllers/BackendController.java:76]: I dont know how spring reacts to endpoints throwing exceptions, I suggest catching the exception in the method, and then returning a response body indicating that something went wrong.

**Review by TotallyRealHolm — CHANGES_REQUESTED** [review]
Code looks very good! :)
I have some requested changed, mostly related to me defining some parts of the API spec in the wrong way, sorry about that.

File: Backend/src/main/java/com/oops/backend/Backend/requests/SolversToUseBody.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    private String MaxMemory;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/requests/SolversToUseBody.java:5]: NumberVCPU Maxmemory and TimeOutPeriod should be int.

File: Backend/src/main/resources/application.properties (hunk 5 above)
Comment (lukasNwr) [reply Backend/src/main/resources/application.properties:1]: I think it is used for writing app specific configuration, e.g port in which the server is going to run... I think I can remove it.

**Review by Maje419 — APPROVED** [review]
LGTM

File: Backend/src/main/java/com/oops/backend/Backend/requests/SolversToUseBody.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
+package com.oops.backend.Backend.requests;
+
+public class SolversToUseBody {
+    private int NumberVCPU;
+    private int MaxMemory;
+    private int TimeOutPeriod;
```
Comment (TotallyRealHolm) [inline Backend/src/main/java/com/oops/backend/Backend/requests/SolversToUseBody.java:6]: This should just be timeout

**Review by TotallyRealHolm — CHANGES_REQUESTED** [review]
Code looks good, few changes in regard to json object schema
